lighthouse-core/audits/page-execution-timings.js

```diff
@@ -0,0 +1,161 @@
+ */
+
+/**
+ * @fileoverview Audit a page to see if it does not use <link> that block first paint.
+ */
+
+'use strict';
+
+const Audit = require('./audit');
+const Util = require('../report/v2/renderer/util');
+const DevtoolsTimelineModel = require('../lib/traces/devtools-timeline-model');
+
+const timelineCategories = [
+  {
+    name: 'Compile Script',
+    group: 'JavaScript',
+  },
+  {
+    name: 'Evaluate Script',
+    group: 'JavaScript',
+  },
+  {
+    name: 'Run Microtasks',
+    group: 'JavaScript',
+  },
+  {
+    name: 'Minor GC',
+    group: 'JavaScript',
+  },
+  {
+    name: 'Major GC',
+    group: 'JavaScript',
+  },
+  {
+    name: 'XHR Ready State Change',
+    group: 'JavaScript',
+  },
+  {
+    name: 'XHR Load',
+    group: 'JavaScript',
+  },
+  {
+    name: 'Layout',
+    group: 'Paint/Layout',
+  },
+  {
+    name: 'Paint',
+    group: 'Paint/Layout',
+  },
+  {
+    name: 'Composite Layers',
+    group: 'Paint/Layout',
+  },
+  {
+    name: 'Update Layer Tree',
+    group: 'Paint/Layout',
+  },
+  {
+    name: 'Recalculate Style',
+    group: 'Paint/Layout',
+  },
+  {
+    name: 'Parse HTML',
+    group: 'DOM/CSS',
+  },
+  {
+    name: 'Parse Stylesheet',
+    group: 'DOM/CSS',
+  },
+  {
+    name: 'DOM GC',
+    group: 'DOM/CSS',
+  },
+  {
+    name: 'Image Decode',
+    group: 'Images',
+  },
+];
+
+class PageExecutionTimings extends Audit {
+
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Performance',
+      name: 'page-execution-timings',
+      description: 'Page executiontime is high',
+      informative: true,
+      helpText: 'Consider reducing the time spent parsing, compiling and executing JS.' +
+        'You may find delivering smaller JS payloads helps with this.',
+      requiredArtifacts: ['traces'],
+    };
+  }
+
+  /**
+   * @param {!Array<TraceEvent>=} trace
+   * @param {!WebInspector.TimelineProfileTree.Node} A grouped and sorted tree
+   */
+  static getTimingsByCategory(trace) {
+    const timelineModel = new DevtoolsTimelineModel(trace);
+
+    return timelineModel.bottomUpGroupBy('EventName');
+  }
+
+  /**
+   * @param {!Array<TraceEvent>=} trace
+   * @return {!Map<string, Number>}
+   */
+  static getExecutionTimingsByCategory(trace) {
+    const bottomUpByName = PageExecutionTimings.getTimingsByCategory(trace);
+
+    const result = new Map();
+    bottomUpByName.children.forEach((value, key) =>
+      result.set(key, value.selfTime.toFixed(1)));
+
+    return result;
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static audit(artifacts) {
+    const trace = artifacts.traces[PageExecutionTimings.DEFAULT_PASS].traceEvents;
+    const executionTimings = PageExecutionTimings.getExecutionTimingsByCategory(trace);
+    let totalExecutionTime = 0;
+
+    const results = timelineCategories.map(category => {
```

Comment:
feels better to loop over your real results instead of iterating the lookup.
with the categories i shared above.. you have constant time lookup:

```js
Array.from(executionTimings).map(([eventName, duration]) => {
    totalExecutionTime += duration;
    return { 
      category: eventName,
      group: taskToGroup[eventName],
      duration: Util.formatMilliseconds(duration, 1),
    };
});
```

Reply:
nice! much better actually 👍